### DISCUSSION TITLE
Why can't I soft navigate to a dynamic route, even with prefetching? (with app directory)

### DISCUSSION THREAD

**Discussion by londonadagio** [vercel/next.js#47296] [Help]
### Summary

Hey everyone !

I've been building an app with the app directory enabled, and I'm struggling to understand prefetching and soft vs hard navigation. 

According to the [documentation](https://beta.nextjs.org/docs/routing/linking-and-navigating#conditions-for-soft-navigation), soft navigation occurs if 1) the link is prefetched and 2) it doesn't point to a dynamic route.
From my testing it behaves exactly like this. If I'm on a static page (say I'm on the home page of my website), and I put a link to a dynamic route (maybe "/[segment]/something"), even with prefetching enabled and in production build, the page is only generated and cached when I click on the link the first time. 

What I'm wondering is, why can't I tell NextJS to prefetch and cache that page so that it can use soft navigation and load instantly? It might not make sense to do it by default, because this could dramatically increase server load if you have a lot of links, and there are pages that you don't want to prefetch at all because you want the most updated data, but I can definitely think of a few instances where I'd want this behavior. But apparently I can't?

The main reason I'm wondering this is because in my experience, dynamic routes are often pages that will have to do some form of API/database call, which can take some time. Those are the pages that could really use prefetching to make navigation faster. Whereas static pages that are prefetched and cached are usually statically generated page that are very fast to fetch to begin with. So right now, in my case at least, soft navigation is only available for the pages that are fast to load anyway, and not available for pages that are slow to load. It feels like it's a bit pointless?

And a related question: is there a point in prefetching dynamic routes at all? Would it be preferable to disable prefetching for those routes, or does that make no difference?

**Comment by rodlac** [comment]
I believe the problem I'm facing maybe could be related to this.  Is there a way to force pages with Dynamic Segments to be cached.  But somehow Vercel's cache servers don't know how to deal with it.

 My problem was reported here: #47406

**Comment by londonadagio** [comment]
I don't think it's the same issue. All the tests I've done were done on a local production build, no Vercel involved.

And from my understanding, when NextJS prefetches a page, it caches it on the client side. My problem is that when it prefetches a dynamic page the server only answers with something like that : 
`0:[["children","post",["post",{"children":[["id","1","d"],{"children":["__PAGE__?{\"id\":\"1\"}",{}]}]}],null,null]]`
instead of the full page. So the page isn't in the client cache, and it needs to be completely fetched when navigating.

But I'll keep an eye on your issue to see if anything comes out of it.

**Reply by rodlac** [reply]
It's not the same, but I believe my problem is related to it, since if the server responded with the complete page in prefetch, there would be no problem with cache on the server and also on the client side. It would solve both problems.

In the [comment I made](https://github.com/vercel/next.js/issues/47406#issuecomment-1479922955_) on my issue, I ran a test forcing the server to cache only the response of the complete page, but this caused another type of problem. However, you can test it and I believe that navigation will partially work as you would expect.

1. Go to [link](https://prefetch-cache-bug-git-no-cache-prefetch-rodlac.vercel.app/).
2. Click on the link of the first post.
3. Go back to the Home.
4. Refresh the page.
5. Click again on the link of the first post.

After refreshing the page in step 4, you will notice that prefetch of the first post will respond with the complete page and soft navigation will work from then on. For testing purposes, this cache is set to only 30 seconds.

**Reply by londonadagio** [reply]
Interesting, I'll try it out.

**Comment by rodlac** [comment]
I think I found a solution to our problem. By adding the generateStaticParams function in page.tsx I got the expected result.

```
export async function generateStaticParams() {
  /*
  // Use this if you want to prefetch a post at build time
  return [{id: '1'}]
  */
  return [];
}
```

You can check from [this link](https://prefetch-cache-bug-git-generate-static-params-rodlac.vercel.app/) that it now works as expected. In [repository](https://github.com/rodlac/prefetch-cache-bug/blob/generate-static-params/app/post/%5Bid%5D/page.tsx) you can see the complete example.

Check if this also solves your problem.

**Reply by londonadagio** [reply]
Thanks for the info, I'll try that !
